### ISSUE TITLE
Benchmark FP16 vs FP32 training on a large-net workload

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#272]
Child of epic #267 (Mixed-precision FP16 training). Phase 3 of the candidate breakdown.

## Context
The feasibility spike (`docs/FP16_FEASIBILITY.md`, PR for #267) measured a real
but modest f16 speedup on Apple Metal (1.9× at a 256×256 MLP, ~4% at 1024×1024 —
memory-bandwidth-bound, no tensor cores). NVIDIA cuda should show a larger f16
win via tensor cores, but this was not measurable on the spike host (macOS).

## Scope
Once the opt-in FP16 path exists (#270), quantify it against the fp32 baseline:
- Extend `benches/trainer_throughput.rs` (or a large-net variant) to run at least
  one benchmark group with the backend swapped to f16, on a **large-net** config —
  the small 4-input/64-unit MLPs will not show a payoff.
- Report paired f16-vs-f32 wall-clock on wgpu **and** on cuda (Linux+NVIDIA host),
  alongside the NdArray CPU reference, matching the methodology in
  `docs/BURN_BACKENDS.md`.
- Update `docs/BURN_BACKENDS.md` / `docs/FP16_FEASIBILITY.md` with the numbers.

## Prerequisite / sequencing
Depends on #270 (opt-in FP16 path) and on a large-net workload existing. Sequenced
behind the same re-evaluation trigger as the rest of epic #267.

## Acceptance criteria
- [ ] Paired f16-vs-f32 bench numbers on a large-net workload (wgpu + cuda)
- [ ] CPU NdArray f32 reference included for context
- [ ] Docs updated with the measured deltas and a clear recommendation

Refs #267.

**Comment by rjwalters** [comment]
Sweep triage: blocking — this benchmark requires the opt-in FP16/bf16 training path from #270, which is now loom:blocked on a real platform gap (cubecl 0.2 has no BF16 matmul kernel on Metal; see the evidence comment on #270). Unblocks when #270 resolves via one of its re-scope options (CUDA-only, f16+readback fix, or a Burn/cubecl version bump).

**Comment by rjwalters** [comment]
Operator review 2026-07-06: stays blocked, chained to #270's sharpened trigger (Burn/cubecl bf16-Metal kernel **or** first large-net workload, whichever first). When #270 executes its option-A path (CUDA-gated, verified on alc-2), this benchmark runs in the same session on the same host — the paired f16-vs-f32 cuda numbers this issue wants are exactly what that verification produces. No independent unblock condition.

**Comment by rjwalters** [comment]
Reframe note: platform tracking (Metal bf16) moved to #305. This issue remains chained to #270, which is now gated solely on the large-net workload trigger — CUDA support already exists, so no upstream movement is required to execute.

**Comment by rjwalters** [comment]
Unblock path now has a concrete vehicle: epic #306 (Atari/ALE env + Nature-DQN-scale CNN) is the large-net workload this issue's trigger names. Phase 4 of that epic explicitly re-triages this issue with measured evidence.

**Comment by rjwalters** [comment]
## Re-triage — workload trigger fired; still gated on #270 execution (Epic #306 Phase 4, 2026-07-09)

The large-net workload that this f16-vs-f32 benchmark needs now exists and is measured (Nature-DQN CNN on RTX 4090). Evidence — see the #270 re-triage and the measured cuda column in `docs/BURN_BACKENDS.md` (PR #344): `nature_dqn_*_train_step/cuda` runs at 4–10 ms vs 954 ms–7.58 s on the alc-2 CPU (213–1230×), so the FP32 baseline for a paired f16-vs-f32 comparison is in hand and clearly compute-bound at this scale.

**Path forward:** this issue stays chained to #270 per the prior operator decision. When #270 executes its CUDA-gated option A on alc-2, the paired f16-vs-f32 cuda numbers this issue wants are produced in the **same session** (same host, same bench harness, warm autotune) — that is the apples-to-apples comparison, and splitting it across sessions would reintroduce run-to-run variance.

**Label action:** keeping `loom:blocked`. The block reason is now narrower — "waiting on #270 to execute option A on alc-2" rather than "waiting for a large-net workload to exist." Unblocks automatically once #270 runs.

**Comment by rjwalters** [comment]
## Curator enrichment — 2026-07-09

Curating per operator instruction: #270 merged as PR #347 today (feat(train): opt-in f16 mixed-precision DQN path behind training-fp16 flag). All prerequisites are now on `origin/main`. Removing `loom:blocked`; adding `loom:curated`.

---

## Technical context grounded in current code

### What #270 (PR #347) delivered — the foundation

PR #347 added three files to `origin/main`:
- `Cargo.toml`: `training-fp16 = ["burn?/wgpu"]` feature. Compose as `--features "training,cuda,training-fp16"` for the verified CUDA path.
- `examples/games/atari/train_pong_dqn_fp16.rs`: cfg-alias pattern selecting `Autodiff<Cuda<f16,i32>>` (or `Wgpu<f16,i32>`) at compile time; dynamic loss-scaler owned by the example (initial scale 4096, ceiling 16384, halve on overflow, double after 2000 clean steps).
- `src/train/dqn/trainer.rs`: `DQNTrainerBurn::train_step_scaled(rng, loss_scale, forward_fn, target_fn)` — the wrapper that multiplies the loss by `loss_scale`, runs backward, unscales gradients via `unscale_grads`, and returns `(DQNStepStatsBurn, applied: bool)` so the caller can detect overflow and shrink the scale.

### Bench harness: current state (origin/main)

`benches/trainer_throughput.rs` has a twelve-group suite via `register_all<B: AutodiffBackend>`:
- Eight small-MLP groups (A2C, PPO, DQN, SAC on CartPole/Pendulum).
- Four Nature-DQN CNN groups (added in issue #328): `nature_dqn_policy_forward`, `nature_dqn_qnet_forward`, `nature_dqn_policy_train_step`, `nature_dqn_qnet_train_step` — each at batch sizes 32 and 256.
- Every group is suffixed `/{suffix}` (e.g., `nature_dqn_qnet_train_step/cuda`).
- The CNN groups use `B::sync(device)` post-kernel to drain the GPU command queue (added in #340).
- The driver registers f32 backends: always `ndarray`, then `wgpu` behind `#[cfg(feature = "wgpu")]`, then `cuda` behind `#[cfg(feature = "cuda")]`.

**Critically: the bench harness has no f16 registration today.** This issue adds it.

### Type-safety analysis: does `register_all<B>` compile for `B = Autodiff<Cuda<f16,i32>>`?

`NatureDqnBurnPolicy<B>` and `NatureDqnQNetwork<B>` both require only `B: Backend` (confirmed in `src/policy/atari_cnn.rs`). `DQNTrainerBurn<B, Q, O>` requires `B: AutodiffBackend, Q: AutodiffModule<B> + Clone, O: Optimizer<Q, B>` — no `FloatNdArrayElement` or f32-specific element bound. Burn's `FloatElement` trait (the cubecl bound) is implemented for `f16` in `burn-cubecl-0.21.0/src/element.rs`, so `Autodiff<Cuda<f16,i32>>` satisfies `AutodiffBackend`. The four existing CNN bench closures use `Tensor<B, 4>` and standard Burn ops — all element-agnostic. **No src/ changes are expected to be needed.** If a compilation failure occurs (e.g., an unexpected `Into<f64>` coercion on a batch-stat path), document and fix narrowly; do not add broad element-type bounds.

### The `train_step_scaled` decision for bench closures

The four CNN bench closures currently use raw forward/backward (not `DQNTrainerBurn::train_step_scaled`) because `bench_nature_dqn_qnet_train_step` builds a standalone `NatureDqnQNetwork` and calls `.backward()` directly. The f16 registration should follow this same pattern — the bench measures raw network throughput, not trainer overhead including the replay buffer. This is correct and representative: the bottleneck being measured is GPU compute on the conv stack, not the loss-scaling arithmetic (two host-side scalar reads + one `mul_scalar` — microseconds on any backend).

**Decision: do NOT add a `train_step_scaled` variant bench.** Rationale: the scaling overhead is O(1) host-side arithmetic around the same backward graph; it does not differ between f32 and f16 in any way that changes the f16-vs-f32 comparison for the GPU compute question. Adding a scaled bench would create noise without signal. The f16 training correctness (including the scaler) is already validated by the #270 acceptance run; this benchmark's job is the isolated throughput comparison.

### f16 bench registration design

Add a new `#[cfg(all(feature = "cuda", feature = "training-fp16"))]` block in `benches()` after the existing cuda block:

```rust
#[cfg(all(feature = "cuda", feature = "training-fp16"))]
{
    type Fp16Cuda = Autodiff<Cuda<burn::tensor::f16, i32>>;
    register_all::<Fp16Cuda>(c, &default_burn_device::<Fp16Cuda>(), "cuda-f16");
}
```

And for wgpu (compiles but unverified at runtime — Metal has no bf16 matmul and f16 is untested per #305):

```rust
#[cfg(all(feature = "wgpu", not(feature = "cuda"), feature = "training-fp16"))]
{
    type Fp16Wgpu = Autodiff<Wgpu<burn::tensor::f16, i32>>;
    register_all::<Fp16Wgpu>(c, &default_burn_device::<Fp16Wgpu>(), "wgpu-f16");
}
```

This produces groups `nature_dqn_policy_forward/cuda-f16`, `nature_dqn_qnet_train_step/cuda-f16`, etc., sitting side-by-side with the existing `/cuda` groups in one criterion run — exactly the paired f16-vs-f32 comparison.

The small-MLP groups (a2c, ppo, dqn, sac) will also register under `cuda-f16` via `register_all`, which is expected and harmless. Those groups will likely show no f16 win (same kernel-overhead story as f32 at small MLP sizes) and serve as a sanity check. The measurement report should focus discussion on the four `nature_dqn_*` groups.

**The bench must NOT add a new `required-features` list** — `training-fp16` is opt-in and the bench already only compiles when its feature is enabled at the `#[cfg]` call site.

### Warm-autotune protocol (alc-2, CUDA)

Per the `docs/BURN_BACKENDS.md` methodology for CUDA:
1. First pass: `cargo bench --features "training,cuda,training-fp16" --bench trainer_throughput -- --warm-up-time 5 --measurement-time 15` — this run triggers cuBLAS/cuDNN autotuning. **Discard these numbers.**
2. Second pass: same command. **Record these numbers.** Criterion's internal warm-up window (5s per group) combined with a prior full run ensures the autotune cache is warm.

Record the exact command (including feature flags) in the docs update.

### Context data point

The #270 acceptance run showed 132 vs ~100 env-steps/s end-to-end (roughly 1.3×) on the IPC-bound Pong training loop. That is dominated by environment IPC overhead (Python subprocess, frame preprocessing), not GPU compute. An isolated bench — measuring only the DQN Q-network backward pass without env stepping — should show a larger compute-only delta where NVIDIA tensor cores activate. The `docs/BURN_BACKENDS.md` existing cuda numbers showed the small-MLP groups favour CPU by ~3×; the expectation is that the CNN groups at B=32 and B=256 break even or show a GPU win at f32, and that f16 on CUDA extends that win via tensor cores.

---

## Refined acceptance criteria

- [ ] `benches/trainer_throughput.rs`: add a `#[cfg(all(feature = "cuda", feature = "training-fp16"))]` block in `benches()` that calls `register_all::<Autodiff<Cuda<f16,i32>>>(c, &device, "cuda-f16")`, producing at minimum the four `nature_dqn_*/cuda-f16` groups.
- [ ] Optionally add a symmetric `wgpu-f16` block behind `#[cfg(all(feature = "wgpu", not(feature = "cuda"), feature = "training-fp16"))]` — mark in docs as "compiles, runtime unverified on Metal" (per #305).
- [ ] The bench compiles without src/ changes. If any element-type bound blocks compilation, document the specific trait gap and add the minimal fix with a comment referencing this issue.
- [ ] `docs/BURN_BACKENDS.md`: add an "## FP16 benchmarks" subsection with a paired table: f32 vs f16 per group + batch size on cuda (and wgpu if measured), with a delta column. Include the exact command used. Note that the small-MLP groups are included for completeness but the signal is in the `nature_dqn_*` groups.
- [ ] `docs/FP16_FEASIBILITY.md`: add a short "## Measured on CUDA (2026-07)" addendum (2–4 sentences) pointing at the new BURN_BACKENDS.md section. Note whether the tensor-core win materialized as predicted (the spike predicted "larger on NVIDIA than Metal") and cross-reference the exact PR.
- [ ] CPU NdArray f32 reference is already in `docs/BURN_BACKENDS.md` from the existing #219 run — do NOT re-measure; cite as-is.
- [ ] No new example binaries, no changes to src/ policy or trainer logic unless required to resolve a compilation failure.
- [ ] **Scope guard**: if the bench does not compile for `Autodiff<Cuda<f16,i32>>` due to a bound that is genuinely infeasible to fix narrowly, document the failure mode and close the bench scope to wgpu-f16 only (or mark blocked on the specific Burn/cubecl limitation).

---

## Measurement plan

**Local (wgpu/Metal, M3 Ultra — for wgpu-f16 if included):**
```bash
cargo bench --features "training,wgpu,training-fp16" --bench trainer_throughput \
    -- --warm-up-time 5 --measurement-time 15
```
Note: wgpu/Metal f16 is not runtime-verified per #305. If it panics, skip wgpu-f16 in the docs table and note the gap.

**alc-2 (CUDA, RTX 4090 — primary measurement, operator-authorized):**
```bash
# Pass 1: warm autotune — discard
cargo bench --features "training,cuda,training-fp16" --bench trainer_throughput \
    -- --warm-up-time 5 --measurement-time 15

# Pass 2: record these numbers
cargo bench --features "training,cuda,training-fp16" --bench trainer_throughput \
    -- --warm-up-time 5 --measurement-time 15
```
Record criterion's per-group median (not mean) from Pass 2. The existing f32 cuda column from the #219 run is the comparison baseline; if a fresh f32 run on the same session is desired for tighter apples-to-apples comparison (avoids between-session variance), it is acceptable but not required — use `--features "training,cuda"` for the f32-only pass and note in the table which pass each number came from.

---

## Affected Files

- `benches/trainer_throughput.rs` — add f16 backend registration blocks in `benches()`
- `docs/BURN_BACKENDS.md` — add FP16 subsection with paired f16-vs-f32 table
- `docs/FP16_FEASIBILITY.md` — add short measured-on-CUDA addendum

Possibly (only if compilation requires it):
- `src/policy/atari_cnn.rs` — no changes expected; B: Backend bound already covers f16
- `src/train/dqn/trainer.rs` — no changes expected; trainer bounds already element-agnostic

---

## Test Plan

- [ ] Compilation: `cargo check --features "training,cuda,training-fp16" --bench trainer_throughput` on alc-2 (or a host with CUDA) produces no errors.
- [ ] Compilation: `cargo check --features "training,wgpu,training-fp16" --bench trainer_throughput` on the M3 Ultra (local) produces no errors.
- [ ] CI (CPU-only, no f16 feature): `cargo bench --features training --bench trainer_throughput` continues to emit only the `*/ndarray` groups — zero regression.
- [ ] Measurement run: criterion Pass 2 on alc-2 emits `nature_dqn_*/cuda-f16` groups with finite, stable medians (no NaN/inf — if those appear, the bench closure has an overflow issue separate from the loss scaler; investigate before recording numbers).
- [ ] Docs update: `docs/BURN_BACKENDS.md` paired table includes at minimum `nature_dqn_policy_forward`, `nature_dqn_qnet_forward`, `nature_dqn_policy_train_step`, `nature_dqn_qnet_train_step` at both batch sizes, with f32-vs-f16 columns and a delta column.
- [ ] `docs/FP16_FEASIBILITY.md` addendum correctly references the BURN_BACKENDS.md section and does not contradict the spike predictions.
